### DISCUSSION TITLE
Opening hours: should hours ranges for a day extend past midnight?

### DISCUSSION THREAD

**Discussion by davidhicks** [alltheplaces/alltheplaces#4959] [Q&A]
I've seen that some source opening hours data that lists e.g. "Monday 08:00 to 02:00" and this is happily interpreted by OpeningHours.add_range() as "Mo 08:00-02:00" rather than "Mo 08:00-24:00; Tu 00:00-02:00". Using the OpenStreetMap <time_domain> specification, this is probably an error because "Mo 08:00-**02:00**; Tu **08:00**-22:00" would have the meaning of "Mo 08:00-**24:00**; Tu **08:00**-22:00" as the hours for Tuesday would override the residual hours from Monday between 12:00-02:00.

Reference: https://wiki.openstreetmap.org/wiki/Key:opening_hours/specification#explain:time_domain

Should OpeningHours.add_range() be changed to automatically detect ranges extending beyond midnight, and cut those ranges into two parts? Or should each spider be more careful to avoid adding ranges extending beyond midnight?

**Comment by Cj-Malone** [comment]
I think `Mo 08:00-02:00`, `Mo 08:00-26:00` and `Mo 08:00-24:00; Tu 00:00-02:00` are all technically valid in OSM world. Maybe I prefer reading the latter, but it's not a strong preference from me, and IMO a data consumer should be parsing and displaying it to people, not just blindly showing them the OSM value.

> Should OpeningHours.add_range() be changed to automatically detect ranges extending beyond midnight, and cut those ranges into two parts? Or should each spider be more careful to avoid adding ranges extending beyond midnight?

If a decision is made, it should be in `OpeningHours`, we don't want to unnecessary replicate the same code in every spider.

**Comment by iandees** [comment]
I've always preferred/mapped using the `Mo 08:00-02:00` style in OSM. It matches what I see on the store's opening hours and is easier to read. It also gives data consumers the option to split them apart if they want to. The split option would allow data consumers to join if they want, I suppose, but it's harder than splitting.

I think my preference would be to keep the hours as presented by the business and not go out of our way to split or join ranges together.

**Comment by davidhicks** [comment]
The issue I have is that the OSM <time_domain> specification states:

> "Sa 16:00-03:00; Su 16:00-23:00" is closed on early Sunday morning, which was probably not intended.

ATP is currently exporting hours that match the pattern above, and thus would be incorrectly interpreted by other consumers of OSM opening_hours data.


Sample code:

>        oh = OpeningHours()
>        oh.add_range("Sa", "16:00", "03:00")
>        oh.add_range("Su", "16:00", "23:00")
>        print(oh.as_opening_hours())

Returns result:

> Sa 16:00-03:00; Su 16:00-23:00

Which means to consumers of OSM data: Closes at midnight between Saturday and Sunday, then reopens at 16:00 on Sunday.

**Reply by iandees** [reply]
The opening_hours OSM syntax is bonkers and I don't really want to spend too much time matching it.

That being said, it looks like we can switch the separator from `;` to `,` and match our desired semantics with the OSM syntax.

**Reply by Cj-Malone** [reply]
Agree that the spec is massively complex, but luckily we don't have to parse, only publish, and we only have to support the common bits. Disagree on how much time it's worth it match it, evidently by my recent work no make the output valid.

Like the format or not, the internet has standardised on it [[1](https://github.com/MobilityData/gbfs/blob/master/gbfs.md#hours-and-dates-of-operation)] [[2](https://schema.org/openingHours)]. Politically should support it to further OSM and ATP data adoption. Technically we should support it so standard OSM tools can work with the data.

I like the idea of switching to a `,`, but politically it annoys OSM people [[3](https://github.com/streetcomplete/StreetComplete/issues/305)], and technically it's moving deeper into the spec so less consumers will be able to parse it.

**Reply by brawer** [reply]
Personally I’d vote for switching to `,` to comply with OSM conventions. If this breaks any clients, they can easily fix it on their side. Perhaps ATP should start publishing release notes that would point out such changes to the format?

**Reply by matkoniecz** [reply]
+1 to switch toward `,` - it more often works like expected, see https://wiki.openstreetmap.org/wiki/Key:opening_hours#Closing_after_midnight example - with `;` overwrite it has silly results 

this change should be relatively simple

> The opening_hours OSM syntax is bonkers

In this case I think that OSM adds relatively little extra complexity compared to what exists in real world.

And some opening hours actually have some silly complexity, and various opening hours have different flavours of complexity ending in nightmarish complexity overall.

I guess that inventing own system or own dialect is possible, but I would not expect ending with something noticeably simpler and there is loss of interoperability. ISO 8601 recurring time intervals syntax is not much simpler (or may be actually worse, depending on who and how judges it). I guess that drastically limiting what can be expressed is also possible.

**Reply by matkoniecz** [reply]
> I like the idea of switching to a ,, but politically it annoys OSM people [https://github.com/streetcomplete/StreetComplete/issues/305]

That person was confused by broken German translation, and they were simply wrong about syntax.

**Reply by iandees** [reply]
I've been reluctant to change this because I personally have always mapped opening hours in OSM with semicolons in all situations. I didn't know about the different meaning for the comma separator and the semicolon syntax makes more sense to me.

How does the OSM syntax handle multiple hour ranges per day group?

**Reply by iandees** [reply]
I've always preferred/mapped using the `Mo 08:00-02:00` style in OSM. It matches what I see on the store's opening hours and is easier to read. It also gives data consumers the option to split them apart if they want to. The split option would allow data consumers to join if they want, I suppose, but it's harder than splitting.

I think my preference would be to keep the hours as presented by the business and not go out of our way to split or join ranges together.

**Reply by davidhicks** [reply]
> How does the OSM syntax handle multiple hour ranges per day group?

Only via a comma delimited list of hour ranges such as `Mo 08:00-11:00,13:00-16:00`.

`Mo 08:00-11:00;13:00-16:00` is invalid syntax and `Mo 08:00-11:00;Mo 13:00-16:00` means just `Mo 13:00-16:00` as the last specified rule (separated by the semicolon) overwrites any earlier rule for Monday.

**Reply by iandees** [reply]
Not sure what's going on with the reply timestamps here, sorry.

My question about the comma was mostly to try to understand how the comma-separated list of time ranges in a single day can be discerned from the comma separated list of days. For example, a restaurant open for lunch and dinner late on Monday:

`Mo 11:00-13:00,16:00-02:00, Tu 11:00-02:00`

**Reply by matkoniecz** [reply]
> how the comma-separated list of time ranges in a single day can be discerned from the comma separated list of days
> `Mo 11:00-13:00,16:00-02:00, Tu 11:00-02:00`

Looking at https://openingh.openstreetmap.de/evaluation_tool/?EXP=Mo%2011%3A00-13%3A00%2C16%3A00-02%3A00%2C%20Tu%2011%3A00-02%3A00&lat=48.7769&lon=9.1844&mode=0&DATE=1706762160000 - it seems that parser supporting OSM opening hours is supposed to be smart enough to handle it (presumably by spotting day-of-week code)

**Reply by matkoniecz** [reply]
> I think my preference would be to keep the hours as presented by the business

+1 to that in general, that is why I like this extended hours syntax like `16:00-02:00`

**Reply by davidhicks** [reply]
https://github.com/opening-hours/opening_hours.js/issues/53#issuecomment-53626335 takes the view that a comma would only be used typically in three circumstances:

1. More than one time range per day (not contentious, ATP already does this frequently in countries where siestas are common).
2. Time ranges wrap around midnight (the issue we discuss here).
3. It is desired to add comments against multiple time ranges in the same day, such as `10:00-12:00 open "quiet hours"` to denote a time of day where the location doesn't have music playing in the background. We don't care about this level of complexity at the moment.

**Reply by matkoniecz** [reply]
> would only be used typically in three circumstances

it is 100% fine to use it also in other circumstances, it is more about preferences

`Mo-Fr 10:00-16:00, Sa 10:00-12:00` and `Mo-Fr 10:00-16:00; Sa 10:00-12:00` are both entirely fine

see also https://github.com/streetcomplete/StreetComplete/issues/773

see also https://github.com/Helium314/SCEE/issues/581where new strategy by SC is described (that uses `;` whenever possible but uses `,` where it would result in incorrect override)

**Reply by matkoniecz** [reply]
Would you accept pull request fixing this and starting to produce opening hours in OSM syntax?

**Reply by davidhicks** [reply]
I agree in `Mo-Fr 10:00-16:00, Sa 10:00-12:00` that there is typically no impact whether a comma or semicolon is used. But in `Mo-Fr 10:00-16:00, Fr 10:00-20:00`  (for extended hours on Fridays), the comma and semicolon have different meanings according to the OSM spec. This situation may occur if the source data we parse has `Weekdays: 10am-4pm, Fridays: 10am-8pm (open late)`.

Always happy to review a pull request. I'm not really clear on what we are proposing to change though.

The key thing I'm happy with is ATP's internal representation of opening hours information is seemingly not impacted by any changes discussed in this thread. What this thread discusses is just an output representation of ATP's internal representation, thus we can easily change the output representation at a later time.

**Reply by matkoniecz** [reply]
> Always happy to review a pull request. I'm not really clear on what we are proposing to change though.

I will need to investigate it, but I hope that replacing `;` by `,` in output will solve the problem (is internal representation even ever doing "lets define day X and the override it with another statement about day X"?)

**Reply by davidhicks** [reply]
Always using a comma instead of semicolon looks OK to me per the OSM spec.

The internal representation is essentially just a dictionary of each week day, with a list of time ranges of each week day. Overlap/duplication of time ranges is allowed and gets sorted out prior to rendering to an OSM `opening_hours` compatible string.

**Comment by j9d3it** [comment]
This looks to be fixed in #18509, which means DATA_FORMAT.md should be updated to reflect the change.
